projects/plugins/jetpack/class.jetpack.php

```diff
@@ -775,6 +775,9 @@ function () {
 			add_action( 'shutdown', array( $this, 'push_stats' ) );
 		}
 
+		// After a successful connection.
+		add_action( 'jetpack_site_registered', array( $this, 'site_registered' ) );
```

Comment:
Maybe we could simplify the whole thing if we hooked into a first time connection only. 

And the callback could work like:
```
{
Jetpack::get_requires_user_connection_modules() // new method that only does one thing
Jetpack::get_default_modules() 
array_diff() // should give you the default modules that don't require user connection 
}
```

And then later when the user connection the regular process just takes care of the rest. 

If the new method in the snippet above (`Jetpack::get_requires_user_connection_modules`) respects the existing `get_available_modules` filter too, then I don't think we need to add all of these new parameters in the existing methods and hooks?

Reply:
> Maybe we could simplify the whole thing if we hooked into a first time registration only.
> ...
> And then later when the user connects the regular process just takes care of the rest and reconnections won't be affected.

For now, I'm trying to put all the cards on the table, as explicitly as I can, because things can get tricky if we want to try to make it smarter... Maybe we can, but I'm on the edge of refactoring the whole thing and this could easily turn into a week project (considering breaking everything up, writing tests, etc...)

So once we get the behavior we want, let's think about how we can optimize it. I don't see how we can hook only at the first-time registration.

Here are some things that were in my mind:

* I did not want to touch the logic used to activate sso and redirect in case of error
* I figured the filters to `get_available_modules` are going to be useful

Now that I think it's working, I'm gonna have a second look to see if this can be simpler (without becoming too obscure). Ideas welcome

Reply:
@leogermani I missed this comment but came back to it after #19068 was opened. 

> I don't see how we can hook only at the first-time registration.

Can we just fire a new do_action() in the `Jetpack_Options::get_option( 'unique_connection' );` check [here](https://github.com/Automattic/jetpack/blob/7895f2f1c8116dec0efe4b6aa6dddf0bfa012914/projects/plugins/jetpack/class.jetpack.php#L4896)? Or is it too late? Or maybe we can copy that logic? It doesn't seem too tricky.

Reply:
Another option could be to start sending an indicator whether a new cache site was created as part of the [register response](https://github.com/Automattic/jetpack/blob/7895f2f1c8116dec0efe4b6aa6dddf0bfa012914/projects/packages/connection/src/class-manager.php#L902) from wpcom.

Reply:
There's also an option that signifies how many times the site has disconnected https://github.com/Automattic/jetpack/blob/7895f2f1c8116dec0efe4b6aa6dddf0bfa012914/projects/plugins/jetpack/class.jetpack.php#L3366

Reply:
Also, looking at the disconnect method it seems we don't clear out the `id` option on disconnect, which holds the ID of the wpcom cache site that was generated. Wouldn't be there if it was a new connection.

Reply:
I any case I think that it would be worth knowing when a site is registering for the first time in the client code - not only for this but for all the existing "things" we do on first-time connection today.